fix(ui): make sure that the screen is updated before waiting for input

fredizzimo opened the pull request:
Neovim can in some cases wait for new input and block even if there is a pending redraw. 

The easiest way to reproduce this is to `:set scrolloff=9999` and `set cursorline` when using multigrid. In this case the `win_viewport` message will not be received, since there are still pending updates, which in practice leads to very jerky motion in Neovide, see https://github.com/neovide/neovide/issues/2794. 

The added tests are also testing this case, and all of them fail without modifications here.

The fix is in theory quite simple, just update the screen if `must_redraw` is set to true. But in practice it needs some more tests. `redraw_later` has two different meanings. Once of them is, redraw the screen as soon as possible, in order to scroll for example. The other meaning is to ask for a redraw of the screen after the next key press, to reset some state for example. The second case is blocked by `!need_wait_return && (State & MODE_CMDLINE) == 0)`, which seems to be all that is needed, but that is purely based on tests, so there might be some case missing.

fredizzimo committed:
fix(ui): make sure that the screen is updated before waiting for input

Ensure the screen is fully updated before blocking for input. This did
not always happen before, for example when setting `cursorline
scrolloff=9999`, which lead to jerky movement when using some GUI
applications.

Because of the duality of redraw_later, this can't be done in
command-line or when waiting for "Press ENTER". In many of those cases
the redraw is expected AFTER the key press, while normally it should
update the screen immediately. So, those special cases are excluded.

justinmk commented on src/nvim/state.c:
```diff
@@ -66,9 +66,11 @@ void state_enter(VimState *s)
-      // Duplicate display updating logic in vgetorpeek()
-      if (((State & MODE_INSERT) != 0 || p_lz) && (State & MODE_CMDLINE) == 0
-          && must_redraw != 0 && !need_wait_return) {
+      // Make sure that the screen is fully updated before blocking for new input.
+      // Due to the duality of redraw_later, this can't be done in command line or when we are
+      // waiting for a return key press. In many of those cases the redraw is actually supposed to
+      // happen after the key press, while normally it's used to update the screen immeditely.
```
typo, also can be tightened a bit.

```suggestion
      // Ensure the screen is fully updated before blocking for input. Because of the duality of
      // redraw_later, this can't be done in command-line or when waiting for "Press ENTER".
      // In many of those cases the redraw is actually supposed to happen after the key press,
      // while normally it should update the screen immediately.
```

justinmk commented on src/nvim/state.c:
```diff
@@ -66,9 +66,11 @@ void state_enter(VimState *s)
       key = K_EVENT;
     } else {
-      // Duplicate display updating logic in vgetorpeek()
-      if (((State & MODE_INSERT) != 0 || p_lz) && (State & MODE_CMDLINE) == 0
```
Why was it checking 'lazyredraw' before doing an eager update?

justinmk commented on test/functional/ui/multigrid_spec.lua:
```diff
@@ -2599,4 +2599,256 @@ describe('ext_multigrid', function()
+
+  describe('centered cursorline', function()
+    before_each(function()
+      command('set cursorline')
+      command('set scrolloff=9999')
```
```suggestion
      -- These settings simulate / achieve / force ... #30576
      command('set cursorline scrolloff=9999')
```

justinmk commented:
> `redraw_later` has two different meanings ...

This would be helpful in the docstring for `redraw_later`. 

Also please include your explanation in the commit message.

justinmk reviewed (approved):
Removing special cases and CI passes, that's a good sign.

I compared the test run to master, and didn't find any new `warning: Screen changes were received after the expected state. This indicates indeterminism` warnings.

justinmk commented on test/functional/ui/multigrid_spec.lua:
```diff
@@ -2599,4 +2599,256 @@ describe('ext_multigrid', function()
+  describe('centered cursorline', function()
+    before_each(function()
+      -- These settings simulate / achieve / force ... #30576
+      command('set cursorline scrolloff=9999')
```
just to be clear: I was hoping you would fill in the rest of the comment :)

fredizzimo replied:
Ok, I will check that again

fredizzimo replied:
Ok, it should be better now :)

fredizzimo commented:
I addressed the review comments, but I did not add any comments to `redraw_later`. I think it's better that someone, who actually understands what it does takes care of that.

Better yet, if it was refactored so that no special cases are needed at all. I have no idea how hard that would be though, but it seems like it should be possible to request a redraw after the input is processed rather than before.

bfredl reviewed (approved):
no idea what is going on by 'lazyredraw' either before or after (but that's not important to me, eventually  'lazyredraw' feature should be replaced by "redrawing is just always good enough" feature) , but otherwise looks right to me.

justinmk commented:
I compared the test run to master, and didn't find any new `warning: Screen changes were received after the expected state. This indicates indeterminism` warnings.

After some bake-time, we could put this in 0.10.3 (in case 0.11 gets delayed).

fredizzimo commented:
It would be great to get this backported to the 0.10.x branch. This also fixes a lot of problems with the built in terminal refresh. It's very noticeable now after I have used the 0.10.3 version without this fix for some time. And the problems in 0.10.3 appears even worse than I remember them being in older version.

Before that version, I only noticed it when running git status, where the screen would refresh after up to one second after it should. But now I also see it when just typing in the terminal, the cursor moves, but the text is invisible until after a while.
